From 5a28a8609b1fa96054c208ccfcceb7fdef9c51fa Mon Sep 17 00:00:00 2001
From: WofWca <wofwca@protonmail.com>
Date: Fri, 1 May 2026 18:41:05 +0400
Subject: [PATCH] fix: smash dead bodies' gibs into ground on fall

Instead of having them fly parallel to the ground.
Only seems to work on other players (non-self), unfortunately.
---
 code/game/g_active.c |  1 +
 code/game/g_combat.c | 46 ++++++++++++++++++++++++++++++++++++++++++++
 code/game/g_cvar.h   |  1 +
 code/game/g_local.h  |  2 ++
 4 files changed, 50 insertions(+)

diff --git a/code/game/g_active.c b/code/game/g_active.c
index ea4133ed3..2be4d9489 100644
--- a/code/game/g_active.c
+++ b/code/game/g_active.c
@@ -915,6 +915,7 @@ void ClientThink_real( gentity_t *ent ) {
 	pm.pmove_msec = pmove_msec.integer;
 
 	VectorCopy( client->ps.origin, client->oldOrigin );
+	VectorCopy( client->ps.velocity, client->oldVelocity );
 
 #ifdef MISSIONPACK
 		if (level.intermissionQueued != 0 && g_singlePlayer.integer) {
diff --git a/code/game/g_combat.c b/code/game/g_combat.c
index 7852925b6..467560c5d 100644
--- a/code/game/g_combat.c
+++ b/code/game/g_combat.c
@@ -1282,6 +1282,52 @@ void G_Damage( gentity_t *targ, gentity_t *inflictor, gentity_t *attacker,
 					KnockbackToKnockbackSpeed( -excess ), dir,
 					targ->client->ps.velocity );
 			}
+			if (
+				targ->health <= GIB_HEALTH && g_blood.integer &&
+				!g_oldGibs.integer &&
+				g_gibsOnCollisionInheritPlayerVelocity.value != 0 &&
+				mod == MOD_FALLING &&
+				client
+			) {
+				// Set player velocity to what it was before they hit the ground
+				// so that the gibs inherit this velocity
+				// (see `cg_gibsInheritPlayerVelocity`).
+				// Note that this only seems to work on other players
+				// and not self (probably due to prediction).
+				//
+				// Note that there is a more accurate crash landing
+				// velocity calculation in `PM_CrashLand`,
+				// but for the purposes of throwing gibs around
+				// `oldVelocity` provides good enough accuracy.
+				VectorScale( client->oldVelocity,
+					g_gibsOnCollisionInheritPlayerVelocity.value,
+					client->ps.velocity );
+
+				// However, there is an issue. When the player
+				// is already landed,
+				// applying downwards velocity redirects that velocity
+				// parallel to the ground instead of clipping it (same as
+				// https://github.com/WofWca/quake3-better-gibs-mod/issues/3),
+				// causing the player camera fly off weirdly.
+				// This is caused by the
+				// "don't decrease velocity when going up or down a slope"
+				// part in `bg_pmove.c`.
+				//
+				// To fix that let's just lift the player off of the ground.
+				//
+				// Note that hypothetically the player camera can end up
+				// in a solid: we shouldn't be so crudely manipulating
+				// player position. However this is, again, good enough
+				// since the player is already dead
+				// and this only affects their camera position.
+				if ( client->ps.groundEntityNum != ENTITYNUM_NONE ) {
+					// There is a 0.25 margin in `PM_GroundTrace`,
+					// this value must be higher than that.
+					// But let's increase by 1 for "snapping"
+					// to save network bandwidth (see `SnapVector`).
+					client->ps.origin[2] += 1;
+				}
+			}
 
 			targ->enemy = attacker;
 			targ->die (targ, inflictor, attacker, take, mod);
diff --git a/code/game/g_cvar.h b/code/game/g_cvar.h
index 185db65b0..2e2a5f0c3 100644
--- a/code/game/g_cvar.h
+++ b/code/game/g_cvar.h
@@ -117,6 +117,7 @@ G_CVAR( g_gibsMissileDirectionKnockbackWeight, "g_gibsMissileDirectionKnockbackW
 	"Note that this affects not just the gibs "
 	"but also the camera velocity of the gibbed player." )
 G_CVAR( g_gibsNewEvGibPlayerParmProtocol, "g_gibsNewEvGibPlayerParmProtocol", "1", CVAR_SYSTEMINFO | CVAR_ARCHIVE, NO_TRACK, NULL )
+G_CVAR( g_gibsOnCollisionInheritPlayerVelocity, "g_gibsOnCollisionInheritPlayerVelocity", "1.0", CVAR_ARCHIVE, NO_TRACK, NULL )
 
 G_CVAR( g_podiumDist, "g_podiumDist", "80", NO_FLAGS, NO_TRACK, NULL )
 G_CVAR( g_podiumDrop, "g_podiumDrop", "70", NO_FLAGS, NO_TRACK, NULL )
diff --git a/code/game/g_local.h b/code/game/g_local.h
index c61ceece3..c477414b3 100644
--- a/code/game/g_local.h
+++ b/code/game/g_local.h
@@ -274,6 +274,8 @@ struct gclient_s {
 	int			latched_buttons;
 
 	vec3_t		oldOrigin;
+	// The `ps.velocity` before the latest `Pmove()`.
+	vec3_t		oldVelocity;
 
 	// sum up damage over an entire frame, so
 	// shotgun blasts give a single big kick
-- 
2.55.0

